intel-sgx/ppid-decryption-tool/Enclave/id_enclave.c

```diff
@@ -0,0 +1,252 @@
+    }
+
+    if (p_public_key == NULL || !sgx_is_within_enclave(p_public_key, key_size)) {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    if (p_ide_report == NULL || !sgx_is_within_enclave(p_ide_report, sizeof(*p_ide_report))) {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    if (crypto_suite != PCE_ALG_RSA_OAEP_3072) {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    if (key_size != sizeof(*p_rsa_pub_key)) {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    if ((p_pce_target_info->attributes.flags & SGX_FLAGS_PROVISION_KEY) != SGX_FLAGS_PROVISION_KEY ||
+        (p_pce_target_info->attributes.flags & SGX_FLAGS_DEBUG) != 0)
+    {
+        //PCE must have access to provisioning key
+        //Can't be debug PCE
+        return(SGX_ERROR_INVALID_PARAMETER);
+    }
+
+    g_rsa_key.e[0] = 0x10001;
+    p_rsa_pub_key = (pce_rsaoaep_3072_encrypt_pub_key_t*)p_public_key;
+
+    sgx_status = sgx_create_rsa_key_pair(REF_RSA_OAEP_3072_MOD_SIZE,
+                                         REF_RSA_OAEP_3072_EXP_SIZE,
+                                         (unsigned char*)g_rsa_key.n,
+                                         (unsigned char*)g_rsa_key.d,
+                                         (unsigned char*)g_rsa_key.e,
+                                         (unsigned char*)g_rsa_key.p,
+                                         (unsigned char*)g_rsa_key.q,
+                                         (unsigned char*)g_rsa_key.dmp1,
+                                         (unsigned char*)g_rsa_key.dmq1,
+                                         (unsigned char*)g_rsa_key.iqmp);
+    if (sgx_status != SGX_SUCCESS) {
+        return sgx_status;
+    }
+
+    // PCE wants the key in big endian
+    size_t i;
+    uint8_t* p_temp;
+    p_temp = (uint8_t*)g_rsa_key.e;
+    for (i = 0; i < REF_RSA_OAEP_3072_EXP_SIZE; i++) {
+        p_rsa_pub_key->e[i] = *(p_temp + REF_RSA_OAEP_3072_EXP_SIZE - 1 - i); //create big endian e
+    }
+    p_temp = (uint8_t*)g_rsa_key.n;
+    for (i = 0; i < REF_RSA_OAEP_3072_MOD_SIZE; i++) {
+        p_rsa_pub_key->n[i] = *(p_temp + REF_RSA_OAEP_3072_MOD_SIZE - 1 - i); //create big endian n
+    }
+
+    do {
+        sgx_status = sgx_sha256_init(&sha_handle);
+        if (SGX_SUCCESS != sgx_status)
+            break;
+
+        sgx_status = sgx_sha256_update(&crypto_suite,
+            sizeof(uint8_t),
+            sha_handle);
+        if (SGX_SUCCESS != sgx_status)
+            break;
+        //(MOD followed by e)
+        sgx_status = sgx_sha256_update(p_rsa_pub_key->n,
+            sizeof(p_rsa_pub_key->n),
+            sha_handle);
+        if (SGX_SUCCESS != sgx_status)
+            break;
+        sgx_status = sgx_sha256_update(p_rsa_pub_key->e,
+            sizeof(p_rsa_pub_key->e),
+            sha_handle);
+        if (SGX_SUCCESS != sgx_status)
+            break;
+        // !!! This is the translation of C++ code into C code. The previous code looked like this:
```

Comment:
Will we ever need to use c++ code again? Is it important to leave this comment behind?